src/Nethermind/Nethermind.Config/IProcessExitSource.cs

```diff
@@ -16,13 +17,18 @@ public class ProcessExitSource : IProcessExitSource
     private CancellationTokenSource _cancellationTokenSource = new();
     public int ExitCode { get; set; } = ExitCodes.Ok;
 
+    private TaskCompletionSource _exitResult = new TaskCompletionSource();
+
+    public Task ExitTask => _exitResult.Task;
+
     public CancellationToken Token => _cancellationTokenSource!.Token;
 
     public void Exit(int exitCode)
     {
         if (CancellationTokenExtensions.CancelDisposeAndClear(ref _cancellationTokenSource))
         {
             ExitCode = exitCode;
+            _exitResult.SetResult();
```

Comment:
why a cancellation token is not enough?

Reply:
Cancellation token need the code itself to stop. Problem is, whatever is running `ethereumRunner.Start` does not stop, until `ethereumRunner.StopAsync`, which is blocked by the `WhenAny`. Now, it can also throw `TaskCancellationException`, but somewhere in MSMS catches and hide it and I don't want this to happen again on other code. So, I'd just make sure `StopAsync` is called.